src/main/java/fi/aalto/cs/apluscourses/ui/utils/TreePathCoder.java

```diff
@@ -0,0 +1,23 @@
+package fi.aalto.cs.apluscourses.ui.utils;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.swing.tree.TreePath;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class TreePathCoder<T> {
+  private final ConcurrentMap<TreePath, T> map = new ConcurrentHashMap<>();
+
+  public T code(@Nullable TreePath treePath) {
+    return treePath == null ? emptyCode()
+        : map.computeIfAbsent(treePath, this::codeInternal);
```

Comment:
lambda-phobia :smile:

Reply:
Well, when I was young, there were no lambdas in Java.  We had to actually implement ourselves each and every interface we ever wanted to use.  And then we skied 50 km to school each morning in -30 °C blizzard while wolves were howling.

But the most incredible thing is that, on those days of old, Java was run in a web browser and _it just worked_. On every platform.  And absolutely no lambdas were needed for that.

Reply:
"Can't teach an old dog new tricks" or something like that... :smile: